test/framework/finalizers_helpers.go

```diff
@@ -0,0 +1,168 @@
+		Expect(err).ToNot(HaveOccurred())
+		obj.SetFinalizers([]string{})
+		Expect(helper.Patch(ctx, obj)).To(Succeed())
+	}
+}
+
+func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace string) map[string]*unstructured.Unstructured {
+	graph, err := clusterctlcluster.GetOwnerGraph(ctx, namespace, proxy.GetKubeconfigPath())
+	Expect(err).ToNot(HaveOccurred())
+
+	objsWithFinalizers := map[string]*unstructured.Unstructured{}
+
+	for _, node := range graph {
+		nodeNamespacedName := client.ObjectKey{Namespace: node.Object.Namespace, Name: node.Object.Name}
+		obj := &unstructured.Unstructured{}
+		obj.SetAPIVersion(node.Object.APIVersion)
+		obj.SetKind(node.Object.Kind)
+		err = proxy.GetClient().Get(ctx, nodeNamespacedName, obj)
+		Expect(err).ToNot(HaveOccurred())
+
+		setFinalizers := obj.GetFinalizers()
+
+		if len(setFinalizers) > 0 {
+			objsWithFinalizers[client.ObjectKey{Namespace: node.Object.Namespace, Name: node.Object.Name}.String()] = obj
+		}
+	}
+
+	return objsWithFinalizers
+}
+
+// assertFinalizersExist ensures that current Finalizers match those in the initialObjectsWithFinalizers.
+func assertFinalizersExist(ctx context.Context, proxy ClusterProxy, namespace string, initialObjsWithFinalizers map[string]*unstructured.Unstructured) {
+	Eventually(func() error {
+		var allErrs []error
+		finalObjsWithFinalizers := getObjectsWithFinalizers(ctx, proxy, namespace)
+
+		for objNamespacedName, obj := range initialObjsWithFinalizers {
+			// verify if finalizers for this resource were set on reconcile
+			if _, valid := finalObjsWithFinalizers[objNamespacedName]; !valid {
+				allErrs = append(allErrs, fmt.Errorf("no finalizers set for %s/%s",
+					obj.GetKind(), objNamespacedName))
+				continue
+			}
+
+			// verify if this resource has the appropriate Finalizers set
+			expectedFinalizers, assert := finalizerAssertion[obj.GetKind()]
+			if !assert {
+				continue
+			}
+
+			setFinalizers := finalObjsWithFinalizers[objNamespacedName].GetFinalizers()
+			if !reflect.DeepEqual(expectedFinalizers, setFinalizers) {
+				allErrs = append(allErrs, fmt.Errorf("expected finalizers do not exist for %s/%s: expected: %v, found: %v",
+					obj.GetKind(), objNamespacedName, expectedFinalizers, setFinalizers))
+			}
+		}
+
+		return kerrors.NewAggregate(allErrs)
+	}).WithTimeout(1 * time.Minute).WithPolling(2 * time.Second).Should(Succeed())
+}
+
+// forceMachineDeploymentTopologyReconcile forces reconciliation of the MachineDeployment.
+func forceMachineDeploymentTopologyReconcile(ctx context.Context, cli client.Client, clusterKey types.NamespacedName) {
+	mdList := &clusterv1.MachineDeploymentList{}
+	clientOptions := (&client.ListOptions{}).ApplyOptions([]client.ListOption{
+		client.MatchingLabels{clusterv1.ClusterNameLabel: clusterKey.Name},
+		client.Limit(1),
+	})
+	Expect(cli.List(ctx, mdList, clientOptions)).To(Succeed())
+
+	if _, ok := mdList.Items[0].GetLabels()[clusterv1.ClusterTopologyOwnedLabel]; ok {
```

Comment:
Should we do this for all MDs in the list instead?

Reply:
Also this function should be removed as part of this [issue](https://github.com/kubernetes-sigs/cluster-api/issues/9532)